modules/local-caching-classloader/src/test/java/org/apache/accumulo/classloader/lcc/LocalCachingContextClassLoaderFactoryTest.java

```diff
@@ -0,0 +1,391 @@
+  }
+
+  @Test
+  public void testInitial() throws Exception {
+    // Create a new context definition file in HDFS, but with invalid content
+    assertTrue(fs.mkdirs(new Path("/contextDefs")));
+    final Path initial = new Path("/contextDefs/InitialContextDefinitionFile.json");
+    try (FSDataOutputStream out = fs.create(initial)) {
+      ContextDefinition def = createContextDef("initial", jarAOrigLocation);
+      out.writeBytes(def.toJson());
+    }
+    assertTrue(fs.exists(initial));
+    final URL initialDefUrl = new URL(fs.getUri().toString() + initial.toUri().toString());
+
+    LocalCachingContextClassLoaderFactory factory = new LocalCachingContextClassLoaderFactory();
+    ClassLoader cl = factory.getClassLoader(initialDefUrl.toString());
+    @SuppressWarnings("unchecked")
+    Class<? extends test.Test> clazzA =
+        (Class<? extends test.Test>) cl.loadClass("test.TestObjectA");
+    test.Test a1 = clazzA.getDeclaredConstructor().newInstance();
+    assertEquals("Hello from A", a1.hello());
+  }
+
+  @Test
+  public void testUpdate() throws Exception {
+    // Create a new context definition file in HDFS
+    assertTrue(fs.mkdirs(new Path("/contextDefs")));
+    final Path defFilePath = new Path("/contextDefs/UpdateContextDefinitionFile.json");
+    final ContextDefinition def = createContextDef("update", jarAOrigLocation);
+    try (FSDataOutputStream out = fs.create(defFilePath)) {
+      out.writeBytes(def.toJson());
+    }
+    assertTrue(fs.exists(defFilePath));
+    final URL updateDefUrl = new URL(fs.getUri().toString() + defFilePath.toUri().toString());
+
+    final LocalCachingContextClassLoaderFactory factory =
+        new LocalCachingContextClassLoaderFactory();
+    final ClassLoader cl = factory.getClassLoader(updateDefUrl.toString());
+    @SuppressWarnings("unchecked")
+    Class<? extends test.Test> clazzA =
+        (Class<? extends test.Test>) cl.loadClass("test.TestObjectA");
+    test.Test a1 = clazzA.getDeclaredConstructor().newInstance();
+    assertEquals("Hello from A", a1.hello());
+
+    // Update the contents of the context definition json file
+    fs.delete(defFilePath, false);
+    assertFalse(fs.exists(defFilePath));
+
+    ContextDefinition updateDef = createContextDef("update", jarDOrigLocation);
+    try (FSDataOutputStream out = fs.create(defFilePath)) {
+      out.writeBytes(updateDef.toJson());
+    }
+    assertTrue(fs.exists(defFilePath));
+
+    // wait 2x the monitor interval
+    Thread.sleep(MONITOR_INTERVAL_SECS * 2 * 1000);
+
+    final ClassLoader cl2 = factory.getClassLoader(updateDefUrl.toString());
+    assertThrows(ClassNotFoundException.class, () -> cl2.loadClass("test.TestObjectA"));
+
+    @SuppressWarnings("unchecked")
+    Class<? extends test.Test> clazzD =
+        (Class<? extends test.Test>) cl2.loadClass("test.TestObjectD");
+    test.Test d1 = clazzD.getDeclaredConstructor().newInstance();
+    assertEquals("Hello from D", d1.hello());
+
+  }
+
```

Comment:
Would be good to test an update the replaces the same class name w/ a differ version of the class.